src/methods/properties.js

```diff
@@ -211,21 +211,25 @@ function ensureIsNotFinal(realm: Realm, O: ObjectValue, P: void | PropertyKeyVal
   if (!O.isFinalObject()) {
     return;
   }
-  if (!realm.isInPureScope()) {
-    // We can't continue because this object is already in its final state.
-    AbstractValue.reportIntrospectionError(O, P);
-    throw new FatalError();
-  }
-  // It's not safe to write to this object anymore because it's already
-  // been used in a way that serializes its final state. We can, however,
-  // leak it if we're in pure scope, and continue to emit assignments.
-  Leak.leakValue(realm, O);
-  if (O.isLeakedObject()) {
-    return;
+  if (realm.isInPureScope()) {
+    // It's not safe to write to this object anymore because it's already
+    // been used in a way that serializes its final state. We can, however,
+    // leak it if we're in pure scope, and continue to emit assignments.
+    Leak.leakValue(realm, O);
+    if (O.isLeakedObject()) {
+      return;
+    }
   }
-  // The object was created outside of pure scope so we couldn't leak.
-  // Give up.
-  AbstractValue.reportIntrospectionError(O, P);
+  // We can't continue because this object is already in its final state.
+  let error = new CompilerDiagnostic(
+    "Mutating an object that has been marked as final is not yet supported " +
```

Comment:
"Marked as final" is an implementation detail that is not part of the mental model of someone using Prepack. Consider a message like "Mutating an object with unknown properties, after some of those properties have already been used, is not supported." In the wiki page you can can discuss Object.assign and perhaps give a code snippet illustrating how things can go wrong.

Reply:
Addressed